File: src/editor/filter/FilterAttribute.java
package editor.filter;

import java.util.Arrays;
import java.util.function.Function;
import java.util.stream.Collectors;

import editor.database.card.Card;
import editor.filter.leaf.BinaryFilter;
import editor.filter.leaf.ColorFilter;
import editor.filter.leaf.FilterLeaf;
import editor.filter.leaf.ManaCostFilter;
import editor.filter.leaf.NumberFilter;
import editor.filter.leaf.TextFilter;
import editor.filter.leaf.TypeLineFilter;
import editor.filter.leaf.VariableNumberFilter;
import editor.filter.leaf.options.multi.CardTypeFilter;
import editor.filter.leaf.options.multi.LegalityFilter;
import editor.filter.leaf.options.multi.SubtypeFilter;
import editor.filter.leaf.options.multi.SupertypeFilter;
import editor.filter.leaf.options.multi.TagsFilter;
import editor.filter.leaf.options.single.BlockFilter;
import editor.filter.leaf.options.single.ExpansionFilter;
import editor.filter.leaf.options.single.LayoutFilter;
import editor.filter.leaf.options.single.RarityFilter;

/**
 * This enum represents an attribute by which a card can be filtered.
 *
 * @author Alec Roelke
 */
public enum FilterAttribute
{
    /** Filter by artist name. */
    ARTIST("Artist", (a) -> new TextFilter(a, Card::artist)),
    /** Filter by block. */
    BLOCK("Block", (a) -> new BlockFilter()),
    /** Filter by card number. */
    CARD_NUMBER("Card Number", (a) -> new NumberFilter(a, (c) -> c.number().stream().map((v) -> Double.valueOf(v.replace("--", "0").replaceAll("[\\D]", ""))).collect(Collectors.toList()))),
    /** Filter by card type. */
    CARD_TYPE("Card Type", (a) -> new CardTypeFilter()),
    /** Filter by converted mana cost. */
    CMC("CMC", (a) -> new NumberFilter(a, Card::cmc)),
    /** Filter by color. */
    COLOR("Color", (a) -> new ColorFilter(a, Card::colors)),
    /** Filter by color identity. */
    COLOR_IDENTITY("Color Identity", (a) -> new ColorFilter(a, Card::colorIdentity)),
    /** Filter by expansion. */
    EXPANSION("Expansion", (a) -> new ExpansionFilter()),
    /** Filter by flavor text. */
    FLAVOR_TEXT("Flavor Text", (a) -> new TextFilter(a, Card::normalizedFlavor)),
    /** Filter by format legality. */
    FORMAT_LEGALITY("Format Legality", (a) -> new LegalityFilter()),
    /** Filter by layout. */
    LAYOUT("Layout", (a) -> new LayoutFilter()),
    /** Filter by loyalty. */
    LOYALTY("Loyalty", (a) -> new VariableNumberFilter(a, (Card c) -> c.loyalty().stream().map((l) -> (double)l.value).collect(Collectors.toList()), Card::loyaltyVariable)),
    /** Filter by mana cost. */
    MANA_COST("Mana Cost", (a) -> new ManaCostFilter()),
    /** Filter by name. */
    NAME("Name", (a) -> new TextFilter(a, Card::normalizedName)),
    /** Filter by power. */
    POWER("Power", (a) -> new VariableNumberFilter(a, (c) -> c.power().stream().map((p) -> p.value).collect(Collectors.toList()), Card::powerVariable)),
    /** Filter by printed text. */
    PRINTED_TEXT("Printed Text", (a) -> new TextFilter(a, Card::normalizedPrinted)),
    /** Filter by printed type line. */
    PRINTED_TYPES("Printed Type Line", (a) -> new TextFilter(a, Card::printedTypes)),
    /** Filter by rarity. */
    RARITY("Rarity", (a) -> new RarityFilter()),
    /** Filter by rules (Oracle) text. */
    RULES_TEXT("Rules Text", (a) -> new TextFilter(a, Card::normalizedOracle)),
    /** Filter by subtype. */
    SUBTYPE("Subtype", (a) -> new SubtypeFilter()),
    /** Filter by supertype. */
    SUPERTYPE("Supertype", (a) -> new SupertypeFilter()),
    /** Filter by user-defined tags. */
    TAGS("Tags", (a) -> new TagsFilter()),
    /** Filter by toughness. */
    TOUGHNESS("Toughness", (a) -> new VariableNumberFilter(a, (c) -> c.toughness().stream().map((p) -> p.value).collect(Collectors.toList()), Card::toughnessVariable)),
    /** Filter by type line. */
    TYPE_LINE("Type Line", (a) -> new TypeLineFilter()),

    /** No filter applied. */
    ANY("<Any Card>", (a) -> new BinaryFilter(true)),
    /** Filter all cards. */
    NONE("<No Card>", (a) -> new BinaryFilter(false)),
    /** Filter using one of the predefined filters.  This should not be used to create a filter. */
    DEFAULTS("Defaults", null),
    /** Group of filters.  This should not be used to create a filter. */
    GROUP("Group", null);

    /**
     * Create a new filter that filters the specified attribute.
     *
     * @param attribute attribute to filter by
     * @return a {@link Filter} that filters by the chosen attribute
     */
    public static Filter createFilter(FilterAttribute attribute)
    {
        switch (attribute)
        {
        case GROUP:
            return new FilterGroup();
        default:
            return attribute.create();
        }
    }

    /**
     * Get the attributes that can be filtered.
     *
     * @return an array containing the attributes that can be filtered (all of the values except for
     * {@link #DEFAULTS} and {@link #GROUP}).
     */
    public static FilterAttribute[] filterableValues()
    {
        return Arrays.stream(values()).filter((f) -> f.filter != null).toArray(FilterAttribute[]::new);
    }

    public static FilterAttribute fromString(String attribute)
    {
        return Arrays.stream(values()).filter((a) -> a.toString().equalsIgnoreCase(attribute)).findAny().get();
    }

    /**
     * Label to display when selecting a filter attribute.
     */
    private final String label;
    /**
     * Function for creating a new filter.
     */
    private final Function<FilterAttribute, FilterLeaf<?>> filter;

    /**
     * Create a new FilterAttribute to filter cards by particular attributes.
     *
     * @param l label to display for this FilterAttribute
     * @param f function for creating a filter
     */
    FilterAttribute(String l, Function<FilterAttribute, FilterLeaf<?>> f)
    {
        label = l;
        filter = f;
    }

    /**
     * @return a new {@link FilterLeaf} that filters this attribute.
     */
    public FilterLeaf<?> create()
    {
        return filter.apply(this);
    }

    @Override
    public String toString()
    {
        return label;
    }
}

File: src/main/scala/editor/gui/filter/FilterAttribute.scala
package editor.gui.filter

import _root_.editor.database.attributes.CardAttribute
import _root_.editor.database.attributes.CombatStat
import _root_.editor.database.attributes.Expansion
import _root_.editor.database.attributes.HasMultiOptionsFilter
import _root_.editor.database.attributes.HasOptions
import _root_.editor.database.attributes.HasSingletonOptionsFilter
import _root_.editor.database.attributes.Loyalty
import _root_.editor.database.attributes.ManaCost
import _root_.editor.database.attributes.ManaType
import _root_.editor.database.attributes.Rarity
import _root_.editor.database.attributes.TypeLine
import _root_.editor.database.card.CardLayout
import _root_.editor.filter.leaf._
import _root_.editor.gui.filter.editor._

import scala.reflect.ClassTag

sealed trait FilterAttribute[T, F <: FilterLeaf] {
  def attribute: CardAttribute[T, F]
  def panel(selector: FilterSelectorPanel): FilterEditorPanel[F]

  override def toString = attribute.toString
}

sealed trait TextFilterAttribute { this: FilterAttribute[Seq[String], TextFilter] =>
  override def panel(selector: FilterSelectorPanel) = TextFilterPanel(attribute.filter, selector)
}

sealed trait NumberFilterAttribute { this: FilterAttribute[?, NumberFilter] =>
  override def panel(selector: FilterSelectorPanel) = NumberFilterPanel(attribute.filter, selector)
}

sealed trait ColorFilterAttribute { this: FilterAttribute[Set[ManaType], ColorFilter] =>
  override def panel(selector: FilterSelectorPanel) = ColorFilterPanel(attribute.filter, selector)
}

sealed trait SingletonOptionsFilterAttribute[T <: AnyRef : ClassTag] { this: FilterAttribute[T, SingletonOptionsFilter[T]] =>
  override def attribute: CardAttribute[T, SingletonOptionsFilter[T]] with HasSingletonOptionsFilter[T]
  override def panel(selector: FilterSelectorPanel) = OptionsFilterPanel(attribute.filter, attribute.options, selector)
}

sealed trait MultiOptionsFilterAttribute[T <: AnyRef : ClassTag, F <: MultiOptionsFilter[T] : ClassTag] { this: FilterAttribute[Set[T], F] =>
  override def attribute: CardAttribute[Set[T], F] with HasMultiOptionsFilter[T]
  override def panel(selector: FilterSelectorPanel) = OptionsFilterPanel(attribute.filter, attribute.options, selector)
}

object FilterAttribute {
  case object NameFilter extends FilterAttribute[Seq[String], TextFilter] with TextFilterAttribute {
    override def attribute = CardAttribute.Name
  }

  case object RulesTextFilter extends FilterAttribute[Seq[String], TextFilter] with TextFilterAttribute {
    override def attribute = CardAttribute.RulesText
  }

  case object PrintedTextFilter extends FilterAttribute[Seq[String], TextFilter] with TextFilterAttribute {
    override def attribute = CardAttribute.FlavorText
  }

  case object ManaCostFilter extends FilterAttribute[Seq[ManaCost], ManaCostFilter] {
    override def attribute = CardAttribute.ManaCost
    override def panel(selector: FilterSelectorPanel) = ManaCostFilterPanel(selector)
  }

  case object RealManaValueFilter extends FilterAttribute[Double, NumberFilter] with NumberFilterAttribute {
    override def attribute = CardAttribute.RealManaValue
  }

  case object EffManaValueFilter extends FilterAttribute[Seq[Double], NumberFilter] with NumberFilterAttribute {
    override def attribute = CardAttribute.EffManaValue
  }

  case object ColorsFilter extends FilterAttribute[Set[ManaType], ColorFilter] with ColorFilterAttribute {
    override def attribute = CardAttribute.Colors
  }

  case object ColorIdentityFilter extends FilterAttribute[Set[ManaType], ColorFilter] with ColorFilterAttribute {
    override def attribute = CardAttribute.ColorIdentity
  }

  case object TypeLineFilter extends FilterAttribute[Seq[TypeLine], _root_.editor.filter.leaf.TypeLineFilter] {
    override def attribute = CardAttribute.TypeLine
    override def panel(selector: FilterSelectorPanel) = TypeLineFilterPanel(selector)
  }

  case object PrintedTypesFilter extends FilterAttribute[Seq[String], TextFilter] with TextFilterAttribute {
    override def attribute = CardAttribute.PrintedTypes
  }

  case object CardTypeFilter extends FilterAttribute[Set[String], MultiOptionsFilter[String]] with MultiOptionsFilterAttribute[String, MultiOptionsFilter[String]] {
    override def attribute = CardAttribute.CardType
  }

  case object SubtypeFilter extends FilterAttribute[Set[String], MultiOptionsFilter[String]] with MultiOptionsFilterAttribute[String, MultiOptionsFilter[String]] {
    override def attribute = CardAttribute.Subtype
  }

  case object SupertypeFilter extends FilterAttribute[Set[String], MultiOptionsFilter[String]] with MultiOptionsFilterAttribute[String, MultiOptionsFilter[String]] {
    override def attribute = CardAttribute.Supertype
  }

  case object PowerFilter extends FilterAttribute[Seq[Option[CombatStat]], NumberFilter] with NumberFilterAttribute {
    override def attribute = CardAttribute.Power
  }

  case object ToughnessFilter extends FilterAttribute[Seq[Option[CombatStat]], NumberFilter] with NumberFilterAttribute {
    override def attribute = CardAttribute.Toughness
  }

  case object LoyaltyFilter extends FilterAttribute[Seq[Loyalty], NumberFilter] with NumberFilterAttribute {
    override def attribute = CardAttribute.Loyalty
  }

  case object LayoutFilter extends FilterAttribute[CardLayout, SingletonOptionsFilter[CardLayout]] with SingletonOptionsFilterAttribute[CardLayout] {
    override def attribute = CardAttribute.Layout
  }

  case object ExpansionFilter extends FilterAttribute[Expansion, SingletonOptionsFilter[Expansion]] with SingletonOptionsFilterAttribute[Expansion] {
    override def attribute = CardAttribute.Expansion
  }

  case object BlockFilter extends FilterAttribute[String, SingletonOptionsFilter[String]] with SingletonOptionsFilterAttribute[String] {
    override def attribute = CardAttribute.Block
  }

  case object RarityFilter extends FilterAttribute[Rarity, SingletonOptionsFilter[Rarity]] with SingletonOptionsFilterAttribute[Rarity] {
    override def attribute = CardAttribute.Rarity
  }

  case object ArtistFilter extends FilterAttribute[Seq[String], TextFilter] with TextFilterAttribute {
    override def attribute = CardAttribute.Artist
  }

  case object CardNumberFilter extends FilterAttribute[Seq[String], NumberFilter] with NumberFilterAttribute {
    override def attribute = CardAttribute.CardNumber
  }

  case object LegalInFilter extends FilterAttribute[Set[String], LegalityFilter] {
    override def attribute = CardAttribute.LegalIn
    override def panel(selector: FilterSelectorPanel) = LegalityFilterPanel(attribute.filter, selector)
  }

  case object TagsFilter extends FilterAttribute[Set[String], MultiOptionsFilter[String]] with MultiOptionsFilterAttribute[String, MultiOptionsFilter[String]] {
    override def attribute = CardAttribute.Tags
  }

  case object AnyCardFilter extends FilterAttribute[Unit, BinaryFilter] {
    override def attribute = CardAttribute.AnyCard
    override def panel(selector: FilterSelectorPanel) = BinaryFilterPanel(true)
  }

  case object NoCardFilter extends FilterAttribute[Unit, BinaryFilter] {
    override def attribute = CardAttribute.NoCard
    override def panel(selector: FilterSelectorPanel) = BinaryFilterPanel(false)
  }

  val values: IndexedSeq[FilterAttribute[?, ?]] = IndexedSeq(NameFilter, RulesTextFilter, PrintedTextFilter, ManaCostFilter, RealManaValueFilter, EffManaValueFilter, ColorsFilter, ColorIdentityFilter, TypeLineFilter, PrintedTypesFilter, CardTypeFilter, SubtypeFilter, SupertypeFilter, PowerFilter, ToughnessFilter, LoyaltyFilter, LayoutFilter, ExpansionFilter, BlockFilter, RarityFilter, ArtistFilter, CardNumberFilter, LegalInFilter, TagsFilter, AnyCardFilter, NoCardFilter)

  val fromAttribute = values.map((a) => a.attribute -> a).toMap
}